service/cachemdw/is_cached_middleware.go

```diff
@@ -0,0 +1,83 @@
+package cachemdw
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/kava-labs/kava-proxy-service/clients/cache"
+	"github.com/kava-labs/kava-proxy-service/decode"
+)
+
+const (
+	CachedContextKey   = "X-KAVA-PROXY-CACHED"
+	ResponseContextKey = "X-KAVA-PROXY-RESPONSE"
+
+	CacheHeaderKey       = "X-Kava-Proxy-Cache-Status"
+	CacheHitHeaderValue  = "HIT"
+	CacheMissHeaderValue = "MISS"
+)
+
+// IsCachedMiddleware returns kava-proxy-service compatible middleware which works in the following way:
+// - tries to get decoded request from context (previous middleware should set it)
+// - tries to get response from the cache
+//   - if present sets cached response in context, marks as cached in context and forwards to next middleware
+//   - if not present marks as uncached in context and forwards to next middleware
+//
+// - next middleware should check whether request was cached and act accordingly:
+func (c *ServiceCache) IsCachedMiddleware(
+	next http.Handler,
+) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// if cache is not enabled - do nothing and forward to next middleware
+		if !c.cacheEnabled {
+			next.ServeHTTP(w, r)
```

Comment:
trace log that we're skippng this middleware

Reply:
done